foundation-ui/services/MountBundle.js

```diff
@@ -0,0 +1,147 @@
+import React, {PropTypes} from 'react';
+
+import Hooks from './Hooks';
+import ServiceFactory from '../factories/ServiceFactory';
+
+const hooks = new Hooks();
+const METHODS_TO_BIND = [
+  'updateState'
+];
+
+/*
+ * Example usage;
+ * import {Mount, MountService} from 'foundation-ui/services/MountBundle';
+ *
+ * class ServiceDetailPlugin {
+ *   getContents(children, props) {
+ *     return <ServiceDetail {...props} />;
+ *   }
+ *
+ *   initialize() {
+ *     // priority order is from [-Inifinity, Inifinity]
+ *     MountService.on('serviceDetail', this.getContents, 0);
+ *   }
+ * }
+ *
+ * class MyComponent extends React.Component {
+ *   render() {
+ *     return (
+ *       <Mount mountId="serviceDetail">
+ *         // What to show when serviceDetail doesn't mount
+ *         <EmptyPage />
+ *       </Mount>
+ *     );
+ *   }
+ * };
+ */
+class Mount extends React.Component {
+  constructor() {
+    super(...arguments);
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  getChildContext() {
+    // Get context passed down from parent Mount if one exists
+    // and add to it so we can create a render tree chain of nested mountIds.
+    // This should just be an Array of string mountIds. We can use the mountIds
+    // to look up each Mount Component on the page and see debug info about it
+    // like which package inserted the component, what the priorities were for
+    // each component if multiple components were registered for a mountId etc.
+    const {mountChain = []} = this.context || {};
+
+    return {mountChain: [...mountChain, this.props.mountId]};
+  }
+
+  componentWillMount() {
+    this.updateState();
+    // Register for changes in MountService specific to this mountId.
+    // Whenever MountService has a change it will end up invoking updateState
+    // so we can get the newly filtered children.
+    hooks.addAction(this.props.mountId, this.updateState);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // Update mount contents when new props are received
+    this.updateState(nextProps);
+  }
+
+  componentWillUnmount() {
+    hooks.removeAction(this.props.mountId, this.updateState);
+  }
+
+  updateState(props = this.props) {
+    let {children, mountId} = props;
+    let blackList = ['children', ...Object.keys(Mount.propTypes)];
+    // Filter props consumed by Mount to create childProps
+    let childProps = Object.keys(props).filter(function (key) {
+      return !blackList.includes(key);
+    }).reduce(function (memo, key) {
+      memo[key] = props[key];
+
+      return memo;
+    }, {});
+
+    let filteredChildren = hooks.applyFilter(
```

Comment:
@MatApple We should be able to highlight mount points without the mount chain, I don't see a real need here; but I see benefits in having it, to allow us to pass down additional information. Wonder if we could achieve the same using props only, in that case I would vote to not use the context as the context API is still an [experimental feature](https://facebook.github.io/react/docs/context.html) and the data flow isn't as transparent as it is using props.